### PR TITLE
feat(core): change `BTreeSet` to `IntervalsTree` in allocations

### PR DESCRIPTION
**Pull request by grishasobol**
1) In `ActiveProgram` fields `allocations` and `pages_with_data` now have type `IntervalsTree`.
2) Changes benchmarks, weights and charging for `alloc`/`free`/`free_range`, because their internal logic has been changed.
3) Add migration for program storage.

### TAKE ATTENTION
1) Now syscall `alloc` returns error (-1) in case `static_pages == max_pages`. Even if program tries to allocate `0` pages (-1) will be returned in that case. Presently we returns `512` (end of static memory and also our limit) if pages == `0` and all possible memory is static.
2) Because of changes in `ActiveProgram` it must be supported by all APIs including `gear-js` @osipov-mit

### CODE REVIEW

File: gsdk/src/storage.rs [hunk 1]
```diff
@@ -317,13 +317,18 @@ impl Api {
     ) -> Result<GearPages> {
         let mut pages = HashMap::new();
 
-        for page in &program.pages_with_data {
+        for page in program
+            .pages_with_data
+            .inner
+            .iter()
+            .flat_map(|(s, e)| s.0..=e.0)
```
Comment (ByteNacked) [inline gsdk/src/storage.rs:324]: It's not quite clear for me what this 'flat_map' actually does. Perhaps improving the naming (such as 'start' and 'end') or extracting it into a function (as it seems 'flat_map(|(s, e)| s.0..=e.0)' is used in several places) would help.

**Review by ByteNacked — APPROVED** [review]

File: gsdk/src/storage.rs (hunk 1 above)
Comment (grishasobol) [reply gsdk/src/storage.rs]: renamed here https://github.com/gear-tech/gear/pull/3791/commits/7002ef4b4a95b6a5198451d80763bd8db4480526

we use here autogenerated types, so I don't think we need to implement for them such functionality

File: core/src/memory.rs [hunk 2]
```diff
@@ -425,90 +407,81 @@ impl AllocationsContext {
...
-        if interval.end() >= self.max_pages {
-            return Err(AllocError::ProgramAllocOutOfBounds);
+        // If trying to allocate zero pages, then returns heap start page (legacy).
+        if pages == WasmPage::from(0) {
+            return Ok(heap.start());
         }
 
-        if let Ok(extra_grow) = Interval::<WasmPage>::try_from(end_mem_page..=interval.end()) {
-            charge_gas_for_grow(extra_grow.len())?;
+        let interval = self
```
Comment (breathx) [inline core/src/memory.rs:428]: this looks so clean now 👍🏻

**Review by breathx — CHANGES_REQUESTED** [review]
Requesting changes due to important comments and ask to impl migrations module folder

Otherwise, looks awesome 🔥

File: core/src/program.rs [hunk 3]
```diff
@@ -186,6 +185,6 @@ mod tests {
         assert_eq!(program.static_pages(), WasmPagesAmount::from(2));
 
         // Has no allocations because we do not set them in new
-        assert_eq!(program.allocations().len(), 0);
+        assert_eq!(program.allocations().points_amount(), Some(0));
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply core/src/program.rs:188]: added for todo in https://github.com/gear-tech/gear/issues/3879

File: core/src/memory.rs [hunk 4]
```diff
@@ -388,7 +376,7 @@ impl AllocationsContext {
             }
         }
 
-        if let Some(&page) = allocations.last() {
+        if let Some(page) = allocations.start() {
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply core/src/memory.rs:379]: fixed

File: core/src/memory.rs [hunk 5]
```diff
@@ -397,7 +385,7 @@ impl AllocationsContext {
                 });
             }
         }
-        if let Some(&page) = allocations.first() {
+        if let Some(page) = allocations.end() {
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply core/src/memory.rs:388]: fixed

File: gtest/src/manager.rs [hunk 6]
```diff
@@ -1106,9 +1110,10 @@ impl JournalHandler for ExtManager {
                 program
                     .allocations()
                     .difference(&allocations)
-                    .flat_map(|page| page.to_iter())
-                    .for_each(|ref page| {
-                        pages_data.remove(page);
+                    .flat_map(IntervalIterator::from)
+                    .flat_map(|p| p.to_iter())
+                    .for_each(|ref p| {
+                        pages_data.remove(p);
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply gtest/src/manager.rs:1116]: done

File: pallets/gear/src/benchmarking/mod.rs [hunk 7]
```diff
@@ -559,27 +558,39 @@ benchmarks! {
...
+        for _ in 0..(r * API_BENCHMARK_BATCH_SIZE) {
+            mem.grow(&mut store, 1.into()).unwrap();
+        }
+
     }
-    verify {
-        verify_process(res.unwrap());
+
+    mem_grow_per_page {
+        let p in 1 .. 800;
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply pallets/gear/src/benchmarking/mod.rs:585]: calculates from constants now

File: utils/wasm-instrument/src/gas_metering/schedule.rs [hunk 8]
```diff
@@ -286,7 +286,6 @@ impl Default for InstructionWeights {
 
 pub struct SyscallWeights {
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply utils/wasm-instrument/src/gas_metering/schedule.rs:287]: issue https://github.com/gear-tech/gear/issues/3970

**Review by breathx — APPROVED** [review]

**Pull request merged**